daprblog/content/posts/2023/dapr_flomesh/index.md

```diff
@@ -0,0 +1,546 @@
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: curl
+---
+apiVersion: v1
+kind: Service
+metadata:
+  name: curl
+  labels:
+    app: curl
+    service: curl
+spec:
+  ports:
+    - name: http
+      port: 80
+  selector:
+    app: curl
+---
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: curl
+spec:
+  replicas: 1
+  selector:
+    matchLabels:
+      app: curl
+  template:
+    metadata:
+      labels:
+        app: curl
+      annotations:
+        dapr.io/enabled: "true"
+        dapr.io/app-id: "curl"
+        dapr.io/enable-api-logging: "true"
+        dapr.io/log-level: "debug"
+    spec:
+      serviceAccountName: curl
+      containers:
+      - image: curlimages/curl
+        imagePullPolicy: IfNotPresent
+        name: curl
+        command: ["sleep", "365d"]
+EOF
+
+sleep 3
+kubectl wait --for=condition=ready pod -n ${NAMESPACE} --all --timeout=60s
+```
+
+### Exporting Services
+
+```shell
+export NAMESPACE=dapr-test
+for CLUSTER_NAME in cluster-1 cluster-3
+do
+  kubectx k3d-${CLUSTER_NAME}
+  kubectl apply -n $NAMESPACE -f - <<EOF
+apiVersion: flomesh.io/v1alpha1
+kind: ServiceExport
+metadata:
+  name: nodeapp
+spec:
+  serviceAccountName: '*'
+  pathRewrite:
+    from: '^/nodeapp/?'
+    to: '/'
+  rules:
+    - portNumber: 3000
+      path: '/nodeapp'
+      pathType: Prefix
+EOF
+sleep 1
+done
+```
+
+After exporting a service, FSM will automatically create Ingress rules for it. With these rules, the services can be accessed through the Ingress.
+
+```shell
+for CLUSTER_NAME_INDEX in 1 3
+do
+  CLUSTER_NAME=cluster-${CLUSTER_NAME_INDEX}
+  ((PORT=80+CLUSTER_NAME_INDEX))
+  kubectx k3d-${CLUSTER_NAME}
+  echo "Getting service exported in cluster ${CLUSTER_NAME}"
+  echo '-----------------------------------'
+  kubectl get serviceexports.flomesh.io -A
+  echo '-----------------------------------'
+  curl -s "http://${HOST_IP}:${PORT}/ports"
+  echo '-----------------------------------'
+done
+```
+
+## Test
+
+Switch to cluster `cluster-2` and make a request from the `curl` pod to test.
+
+```shell
+kubectx k3d-cluster-2
+curl_client="$(kubectl get pod -n curl -l app=curl -o jsonpath='{.items[0].metadata.name}')"
+```
+
+An error occurred when sending a request to access the nodeapp because the nodeapp application was not deployed in `cluster-2`. By default, when no cross-cluster traffic policy is specified, it will only attempt to call local services and will not schedule traffic to other clusters.
+
+```shell
+kubectl exec "${curl_client}" -n curl -c curl -- curl -s http://nodeapp.dapr-test:3000/ports
+command terminated with exit code 7
+```
+
+### Setting Traffic Policies
+
+There are three types of traffic policies for cross-cluster traffic management:
+
+- `Locality`: Only use local services, which is the default type. This is why accessing the nodeapp application fails when we don't provide any global policies, because the service is not available in the `cluster-2`.
```

Comment:
```suggestion
- `Locality`: Default traffic policy type that only uses local services. Accessing the NodeApp application fails when you don't provide any global policies, because the service is not available in the `cluster-2`.
```